book/invalidation.md

```diff
@@ -173,96 +166,86 @@ Why Invalidation?
-we're going to need re-rendering the page to take time proportional to
-the *size of the change*, and not proportional to the *size of the
-page*. I call this the principle of incremental performance, and it's
-crucial for handling large and complex web applications. It means that
-developers can predict how long their code will run regardless of the
-content on the page, and it lets users predict how long interactions
-will take. That ultimately makes this principle a core part of the
-value that browser engineers provide to the web as a whole.
-
-But the principle of incremental performance also really constrains
-our browser. For example, during rendering even *traversing* the whole
-layout tree would take time proportional to the whole page, not the
-change being made, so we can't do that. To achieve incrementality,
-we're going to need to think of the initial render and later
-re-renders differently.[^big-change] When the page is first loaded,
-rendering is definitely going to take time proportional to the size of
-the page. But we treat that initial render as a cache. Later renders
-will only invalidate and recompute the parts of that cache, taking
-time proportional to the changed portion of the page.
+    techniques from this chapter, but invalidation is still essential
+    for incremental performance, which is a kind of asymptotic
+    guarantee that simple optimizations won't achieve.
+
+We want interactions to be fast, even on large, complex pages, so we
+want re-rendering the page to take time proportional to the *size of
+the change*, and not proportional to the *size of the page*. I call
+this the principle of incremental performance, and it's crucial for
+handling large and complex web applications. Not only does it mean
+fast text editing, it also means that developers can also think about
+performance one change at a time, without consider the contents of the
+whole page. The incremental performance is therefore necessary for
+complex applications. But the principle of incremental performance
+also really constrains our browser. For example, during rendering even
+*traversing* the whole layout tree would take time proportional to the
+whole page, not the change being made, so we can't do that.
+
+To achieve incremental performance, we're going to need to think of
+the initial render and later re-renders differently.[^big-change] When
+the page is first loaded, rendering will take time proportional to the
+size of the page. But we'll treat that initial render as a cache.
+Later renders will only invalidate and recompute parts of that cache,
+taking time proportional to the size of the change.
 
 [^big-change]: While initial and later renders are conceptually
     different, they'll use the same code path. Basically, the initial
     render will be one big change from no page to the initial page,
-    while later re-renders will incorporate smaller changes.
+    while later re-renders will handle smaller changes.
```

Comment:
```suggestion
    while later re-renders will handle smaller changes. But since a
    "small" change could just be setting the `innerHTML` of the whole
    DOM, all renders can also be viewed as equivalent. The key insight
    is to think about *common interaction patterns* we expect users
    and developers to use, and optimize for them.
```